conformance/scenarios/httpgateway/httpgateway_test.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package httpgateway
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/service-apis/conformance/utils"
+)
+
+func TestHttpGateway(t *testing.T) {
+	// Test Setup
+	c, sc, err := utils.LoadClientset()
+	if err != nil {
+		t.Fatalf("error loading clientset: %v", err)
+	}
+	err = utils.DynamicApply(utils.DynamicParams{Path: "../../../config/crd/bases"})
+	if err != nil {
+		t.Fatalf("error installing service-apis CRDs: %v", err)
+	}
+	ns, err := utils.NewNamespace(c)
+	if err != nil {
+		t.Fatalf("error creating namespace: %v", err)
+	}
+
+	controllerName := os.Getenv("CONTROLLER_NAME")
+	if controllerName == "" {
+		t.Fatalf("CONTROLLER_NAME environment variable must be set")
+	}
+
+	dp := utils.DynamicParams{
+		Path:      "httpgateway.yaml",
+		Namespace: ns,
+		Params:    map[string]string{"Controller": controllerName},
+	}
+	err = utils.DynamicApply(dp)
+	if err != nil {
+		t.Fatalf("error installing service-apis examples: %v", err)
+	}
```

Comment:
I think we should simplify this boilerplate. I've a few ideas to do it:
1. Move the install CRDs part to global setup phase that sets up the k8s cluster. Or as part of running the tests. CRDs should be installed only once during the entire test run.
2. We should avoid templating if possible. Please see my comment in guidelines.md.
3. We should even avoid namespace creation in test. We can have a set of namespaces that are created (and then deleted) as part of the testing framework itself. The author of tests can then assume that these namespaces will be present. For example, the testing cluster will contain `test-sapi-0` to `test-sapi-9` namespaces. 
4. Change `c, sc, err := utils.LoadClientset()` to `c, sc := utils.LoadClientset(t)` and perform t.Fatalf inside the function itself. This gets rid off error checking in the test.
5. The simple case, where the YAMLs in the same dir as this test need to be applied to the cluster and then subsequently deleted (deferred) can be two lines:
```go
cleanup := utils.Apply(t, c)
defer cleanup()
```

This gets us to 3 lines of setup code that needs to be repeated in every test. Plus we also get rid off the cleanup phase below.

I might have gone too deep into details here, pardon me if you were looking for feedback from a different perspective.

Reply:
```
cleanup := utils.Apply(t, c)
defer cleanup()
```

can be reduced to just `utils.Apply(t, c)` (or maybe `utils.ApplyAndCleanup(t, c)` if we want to make it optional) if you use t.Cleanup() in the Apply function rather than defer